docs: Clarify how to configure plain login (#1598)
redpanda-data/documentation: modules/develop/pages/produce-data/leader-pinning.adoc

= Leader Pinning
:description: Learn about Leader Pinning and how to configure a preferred partition leader location based on cloud availability zones or regions.

> :page-topic-type: how-to

Reviewer: Note that we'll also need to add these attributes to https://github.com/redpanda-data/cloud-docs/blob/main/modules/develop/pages/produce-data/leader-pinning.adoc cc @micheleRP

:personas: streaming_developer, platform_admin
:learning-objective-1: Configure preferred partition leader placement using rack labels
:learning-objective-2: Configure ordered rack preference for priority-based leader failover
:learning-objective-3: Identify conditions where Leader Pinning cannot place leaders in preferred racks
// tag::single-source[]

> Produce requests that write data to Redpanda topics route through the topic partition leader, which syncs messages across its follower replicas. For a Redpanda cluster deployed across multiple availability zones (AZs), Leader Pinning ensures that a topic's partition leaders are geographically closer to clients, which helps decrease networking costs and guarantees lower latency.

Reviewer 2: ```suggestion
Produce requests that write data to Redpanda topics are routed through the topic partition leader, which syncs messages across its follower replicas. For a Redpanda cluster deployed across multiple availability zones (AZs), Leader Pinning ensures that a topic's partition leaders are geographically closer to clients, which helps decrease networking costs and guarantees lower latency.
```

Reviewer 2: Also, @kbatuigas can you add a glossterm for Leader Pinning and link to it from here? thx

> If consumers are located in the same preferred region or AZ for Leader Pinning, and you have not set up xref:develop:consume-data/follower-fetching.adoc[follower fetching], Leader Pinning can also help reduce networking costs on consume requests.

Reviewer 2: Thinking maybe you should also mention Cloud Topics, as this is another way to reduce networking costs (at a slight expense to latency, but great for logs, data not reliant on speed).

Reviewer: I added a TODO in the Suggested Reading section - that might be a better place to mention Cloud Topics instead of introducing multiple new ideas/contexts right at the start of this doc.

endif::[]

== Limitations

Leader Pinning controls which replica is elected as leader, and does not move replicas to different brokers. If all of a topic's replicas are on brokers in non-preferred racks, there is no replica in the preferred racks to elect as leader, and it is possible to have a non-preferred leader indefinitely.

For example, consider a cluster deployed across four racks (A, B, C, D) with Leader Pinning configured as `ordered_racks:A,B,C,D`. With a replication factor of 3, rack awareness can only place replicas in three of the four racks. If the highest-priority rack (A) does not receive a replica, there is no replica there to elect as leader, and it is possible to have a non-preferred leader indefinitely.

ifndef::env-cloud[]
To prevent this scenario:

* Enable config_ref:enable_rack_awareness,true,properties/cluster-properties[`enable_rack_awareness`] to distribute replicas across racks automatically.
* Set config_ref:default_leaders_preference,true,properties/cluster-properties[`default_leaders_preference`] before creating topics, so that rack-aware replica placement aligns with the preferred racks.

> * Use xref:manage:cluster-maintenance/cluster-balancing.adoc[partition balancing] to move existing replicas to preferred racks.

Reviewer: Is this sufficient or should we link/mention a specific kind of balancing listed on this page? https://deploy-preview-1598--redpanda-docs-preview.netlify.app/current/manage/cluster-maintenance/cluster-balancing/

endif::[]
ifdef::env-cloud[]
To prevent this scenario:

* Configure `redpanda.leaders.preference` before creating topics to help ensure replicas land in preferred racks from the start.
* Ensure the topic's replication factor is no greater than the number of preferred racks.

endif::[]

== Leader Pinning failover across availability zones

If there are three AZs: A, B, and C, and A becomes unavailable, the failover behavior is as follows:

* A topic with “A” as the preferred leader AZ will have its partition leaders uniformly distributed across B and C.
* A topic with “A,B” as the preferred leader AZs will have its partition leaders in B.
* A topic with “B” as the preferred leader AZ will have its partition leaders in B as well.

=== Failover with ordered rack preference

With `ordered_racks`, the failover order follows the configured priority list. Leaders move to the next available rack in the list when higher-priority racks become unavailable.

For a topic configured with `ordered_racks:A,B,C`:

* A topic with "A" as the first-priority rack will have its partition leaders in A.
* If A becomes unavailable, leaders move to B.
* If A and B become unavailable, leaders move to C.

> * If A, B, and C all become unavailable, leaders fall back to any available brokers.

Reviewer: What happens when a preferred rack becomes available again? Do leaders move back?

Reviewer 3: Yeah, so long as the higher priority region again gets a replica.

RF=5, nodes across racks A B C D E
A failure, B failure
leaders will go to C

A returns
RF=5 dictates that we must have a replica on A
then upon recovery of A, leaders will all get moved to A

Reviewer: Updated in line below

include::shared:partial$enterprise-license.adoc[]
====

Before you can enable Leader Pinning, you must xref:manage:rack-awareness.adoc#configure-rack-awareness[configure rack awareness] on the cluster. If the config_ref:enable_rack_awareness,true,properties/cluster-properties[] cluster configuration property is set to `false`, Leader Pinning is disabled across the cluster.

endif::[]

ifndef::env-cloud[]
== Set leader rack preferences

> You can use both a topic configuration property and a cluster configuration property to configure Leader Pinning.

Reviewer 2: Is this an either/or? You can configure Leader Pinning using _either_ a topic configuration property or cluster configuration property? Or, do you have to configure _both_ the topic and cluster configs?

Reviewer 2: Why is the explanation for this sentence in a separate paragraph? Maybe keep together so people won't ask the same questions I asked above?

Reviewer: Updated

You can set the topic configuration property for individual topics only, or set the cluster-wide configuration property that enables Leader Pinning by default for all topics. You can also use a combination in which a default setting applies across the cluster, and you toggle the setting on or off for specific topics.

This configuration is based on the following scenario: you have Redpanda deployed in a multi-AZ or multi-region cluster, and you have configured each broker so that the config_ref:rack,true,properties/broker-properties[] configuration property contains racks corresponding to the AZs:

> * Set the topic configuration property xref:reference:properties/topic-properties.adoc#redpanda-leaders-preference[`redpanda.leaders.preference`]. The property accepts the following string values:

Reviewer 2: ```suggestion
* Set the topic configuration property xref:reference:properties/topic-properties.adoc#redpanda-leaders-preference[`redpanda.leaders.preference`]. This property accepts the following string values:
```

+
--

> ** `none`: Opt out the topic from Leader Pinning.

Reviewer 4: "Opt the topic out of leader pinning"?

Reviewer: Change to "Disable Leader Pinning for the topic"

** `racks:<rack1>[,<rack2>,...]`: Specify the preferred location (rack) of all topic partition leaders. The list can contain one or more racks, and you can list the racks in any order. Spaces in the list are ignored, for example: `racks:rack1,rack2` and `racks: rack1,  rack2` are equivalent. You cannot specify empty racks, for example: `racks: rack1,,rack2`. If you specify multiple racks, Redpanda tries to distribute the partition leader locations equally across brokers in these racks.
** `ordered_racks:<rack1>[,<rack2>,...]`: (Supported in Redpanda version 26.1 or later) Specify the preferred racks in priority order. Redpanda places leaders in the first listed rack when available, failing over to each subsequent rack when higher-priority racks are unavailable. If all listed racks are unavailable, leaders fall back to any other available brokers. Brokers with no rack assignment are treated as lowest priority. 
+
Use `ordered_racks` for multi-region deployments with a primary region for leaders and explicit failover to a disaster recovery site.
--
+
This property inherits the default value from the cluster property `default_leaders_preference`.
+

> To find the rack identifier, run:

Reviewer 4: The rack identifier of what?  You mean "To get the rack identifiers of all brokers"?  I think this is what you get with that command

Reviewer: Updated

+
[,bash]
----
rpk cluster info
----
+
To set the topic property:
+
[,bash]
----
rpk topic alter-config <topic-name> --set redpanda.leaders.preference=ordered_racks:<rack1>,<rack2>

> ----

Reviewer 2: Since this is part of the topic config, shouldn't it all be indented under that bullet?

<img width="973" height="908" alt="Image" src="https://github.com/user-attachments/assets/cad8be9a-365b-4ad3-a1f2-5ee39ea3bacc" />

Reviewer 2: Here's appears it looks in Cloud, which looks correct:

<img width="970" height="402" alt="Image" src="https://github.com/user-attachments/assets/5ae309a1-5ab8-46ec-a886-41295fde4974" />

Reviewer: The rpk commands are not part of the preceding list items (config property values) and instead are logically connected to the parent list item (setting the topic property), so they should align with the first bullet point instead. Updating the Cloud version to match the Self-managed page, which is correct

* Set the cluster configuration property config_ref:default_leaders_preference,true,properties/cluster-properties[], which specifies the default Leader Pinning configuration for all topics that don’t have `redpanda.leaders.preference` explicitly set. It accepts values in the same format as `redpanda.leaders.preference`. Default: `none`
+
This property also affects internal topics, such as `__consumer_offsets` and transaction coordinators. All offset tracking and transaction coordination requests get placed within the preferred regions or AZs for all clients, so you see end-to-end latency and networking cost benefits.
+
To set the cluster property:
+
[,bash]
----

> rpk cluster config set default_leaders_preference ordered_racks:<rack1>,<rack2>

Reviewer 2: hmmm, why is this set/formatted differently from 
```
rpk topic alter-config <topic-name> --set redpanda.leaders.preference=ordered_racks:<rack1>,<rack2>?

rpk cluster config set default_leaders_preference ordered_racks:<rack1>,<rack2>
```

Reviewer: `rpk topic alter-config --set` requires the `key=value` notation but `rpk cluster config set` can take either `key value` (if setting a single property) or `key=value`, so these examples are correct.

----

If there is more than one broker in the preferred AZ (or AZs), Leader Pinning distributes partition leaders uniformly across brokers in the AZ.

endif::[]

ifdef::env-cloud[]

> == Set leader rack preferences

Reviewer 2: FYI--the learning objectives are not rendering in the Cloud output: 

<img width="352" height="172" alt="Image" src="https://github.com/user-attachments/assets/a15659c9-4fee-4abd-9607-f9a1c10eb4af" />

Reviewer: Updated so that the single source tag includes the learning objectives attributes

Configure Leader Pinning if you have Redpanda deployed in a multi-AZ or multi-region cluster and your ingress is concentrated in a particular AZ or region.

Use the topic configuration property `redpanda.leaders.preference` to configure Leader Pinning for individual topics. The property accepts the following string values:

> ** `none`: Opt out the topic from Leader Pinning.

Reviewer 4: "Opt the topic out of leader pinning"?

Reviewer: Changed to "Disable Leader Pinning for the topic"

For example, consider a cluster deployed across four racks (A, B, C, D) with Leader Pinning configured as `ordered_racks:A,B,C,D`. With a replication factor of 3, rack awareness can only place replicas in three of the four racks. If the highest-priority rack (A) does not receive a replica, no replica exists there to elect as leader, and Redpanda may elect a non-preferred leader indefinitely.

ifndef::env-cloud[]
To prevent this scenario:

* Enable config_ref:enable_rack_awareness,true,properties/cluster-properties[`enable_rack_awareness`] to distribute replicas across racks automatically.
* Ensure the topic's replication factor at least equals the total number of racks in the cluster, so every rack, including the highest-priority rack, receives a replica.
* If needed, manually reassign replicas to ensure the highest-priority rack receives one. Note that the partition balancer may move replicas again after manual reassignment.

endif::[]
ifdef::env-cloud[]

> To prevent this scenario, ensure the topic's replication factor at least equals the total number of racks in the cluster, so every rack, including the highest-priority rack, receives a replica.

Reviewer 4: Seems like you just said this above

Reviewer: This is a conditionalized section, so this is what displays for Cloud docs (and leaves out the mention of `enable_rack_awareness` and manual reassignment)

endif::[]

== Leader Pinning failover across availability zones

If there are three AZs: A, B, and C, and A becomes unavailable, the failover behavior with `racks` is as follows:

* A topic with `A` as the preferred leader AZ will have its partition leaders uniformly distributed across B and C.
* A topic with `A,B` as the preferred leader AZs will have its partition leaders in B.

> * A topic with `B` as the preferred leader AZ will have its partition leaders in B as well.

Reviewer 2: ```suggestion
* The topic with `A` as the preferred leader AZ will have its partition leaders uniformly distributed across B and C.
* The topic with `A,B` as the preferred leader AZs will have its partition leaders in B.
* The topic with `B` as the preferred leader AZ will have its partition leaders in B as well.
```

Reviewer 2: At first it was confusing to see the init "A"...especially since it is also the name of one of the AZs.

=== Failover with ordered rack preference

With `ordered_racks`, the failover order follows the configured priority list. Leaders move to the next available rack in the list when higher-priority racks become unavailable.

For a topic configured with `ordered_racks:A,B,C`:

> * A topic with `A` as the first-priority rack will have its partition leaders in A.

Reviewer 2: ```suggestion
* The topic with `A` as the first-priority rack will have its partition leaders in A.
```

* If A becomes unavailable, leaders move to B.
* If A and B become unavailable, leaders move to C.
* If A, B, and C all become unavailable, leaders fall back to any available brokers.

If a higher-priority rack recovers and the topic's replication factor ensures that rack receives a replica, Redpanda automatically moves leaders back to the highest available preferred rack.

> == Suggested reading

Reviewer 2: Add Cloud Topics?

Reviewer: Added TODO. This should be updated when the Cloud Topics doc becomes available, otherwise it will result in build errors.

+
[,bash]
----
rpk topic alter-config <topic-name> --set redpanda.leaders.preference=ordered_racks:<rack1>,<rack2>
----

If there is more than one broker in the preferred AZ (or AZs), Leader Pinning distributes partition leaders uniformly across brokers in the AZ.

endif::[]

== Limitations

Leader Pinning controls which replica is elected as leader, and does not move replicas to different brokers. If all of a topic's replicas are on brokers in non-preferred racks, no replica exists in the preferred racks to elect as leader, and Redpanda may elect a non-preferred leader indefinitely.

For example, consider a cluster deployed across four racks (A, B, C, D) with Leader Pinning configured as `ordered_racks:A,B,C,D`. With a replication factor of 3, rack awareness can only place replicas in three of the four racks. If the highest-priority rack (A) does not receive a replica, no replica exists there to elect as leader, and Redpanda may elect a non-preferred leader indefinitely.

ifndef::env-cloud[]
To prevent this scenario:

* Enable config_ref:enable_rack_awareness,true,properties/cluster-properties[`enable_rack_awareness`] to distribute replicas across racks automatically.

> * Set config_ref:default_leaders_preference,true,properties/cluster-properties[`default_leaders_preference`] before creating topics, so that rack-aware replica placement aligns with the preferred racks.

Reviewer 3: This is not correct. The default leader preference will not influence the replica placement. As of right now there is no automatic solution for this scenario.

It is possible to place the replicas manually, though there is no guarantee that a balancer action will not move it later.

Reviewer: @joe-redpanda Updated - is the manual reassignment worth mentioning, like so? https://github.com/redpanda-data/docs/pull/1598/changes#diff-09b2e0ca15e9a9e4ce3b4f194471533ebd3fcb2069c7393dac28d6e138cc5a2fR121-R122 And is it done using [rpk cluster partitions move](https://deploy-preview-1596--redpanda-docs-preview.netlify.app/current/reference/rpk/rpk-cluster/rpk-cluster-partitions-move/)?

Reviewer 3: Rack constraint repair only works with the autobalancer set to continuous, which means any manual movements are probably futile because the balancer may at any point decide to shuffle the replicas.

Reviewer: @joe-redpanda The bullet point has been removed, thanks!

* Use xref:manage:cluster-maintenance/cluster-balancing.adoc[partition balancing] to move existing replicas to preferred racks.
* Ensure the topic's replication factor is no greater than the number of preferred racks.

endif::[]
ifdef::env-cloud[]
To prevent this scenario:

> * Configure `redpanda.leaders.preference` before creating topics to help ensure replicas land in preferred racks from the start.

Reviewer 3: This is not accurate. leaders preference will not influence where the replicas land

Reviewer: Removed

> * Ensure the topic's replication factor is no greater than the number of preferred racks.

Reviewer 3: The replication factor being larger than the preferred racks is totally fine

If the replication factor is less than the total number of racks is when it becomes possible for indefinite suboptimal leadership pinning

Reviewer: Updated

This configuration is based on the following scenario: you have Redpanda deployed in a multi-AZ or multi-region cluster, and you have configured each broker so that the config_ref:rack,true,properties/broker-properties[] configuration property contains racks corresponding to the AZs:

* Set the topic configuration property xref:reference:properties/topic-properties.adoc#redpanda-leaders-preference[`redpanda.leaders.preference`]. This property accepts the following string values:
+
--
** `none`: Disable Leader Pinning for the topic.
** `racks:<rack1>[,<rack2>,...]`: Specify the preferred location (rack) of all topic partition leaders. The list can contain one or more racks, and you can list the racks in any order. Spaces in the list are ignored, for example: `racks:rack1,rack2` and `racks: rack1,  rack2` are equivalent. You cannot specify empty racks, for example: `racks: rack1,,rack2`. If you specify multiple racks, Redpanda tries to distribute the partition leader locations equally across brokers in these racks.

> ** `ordered_racks:<rack1>[,<rack2>,...]`: (Supported in Redpanda version 26.1 or later) Specify the preferred racks in priority order. Redpanda places leaders in the first listed rack when available, failing over to each subsequent rack when higher-priority racks are unavailable. If all listed racks are unavailable, leaders fall back to any other available brokers. Brokers with no rack assignment are treated as lowest priority.

Reviewer 2: ```suggestion
** `ordered_racks:<rack1>[,<rack2>,...]`: Supported in Redpanda v26.1 or later. Specify the preferred racks in priority order. Redpanda places leaders in the first listed rack when available, failing over to each subsequent rack when higher-priority racks are unavailable. If all listed racks are unavailable, leaders fall back to any other available brokers. Brokers with no rack assignment are treated as lowest priority.
```

+
Use `ordered_racks` for multi-region deployments with a primary region for leaders and explicit failover to a disaster recovery site.

> This property inherits the default value from the cluster property `default_leaders_preference`.

Reviewer 2: ```suggestion
The xref:reference:properties/topic-properties.adoc#redpanda-leaders-preference[`redpanda.leaders.preference`] property inherits the default value from the cluster property `default_leaders_preference`.
```

To find the rack identifiers of all brokers, run:

> [,bash]

Reviewer 2: can you show the output?

Reviewer: added

----
rpk cluster info
----

To set the topic property:

[,bash]
----
rpk topic alter-config <topic-name> --set redpanda.leaders.preference=ordered_racks:<rack1>,<rack2>
----
--

> * Set the cluster configuration property config_ref:default_leaders_preference,true,properties/cluster-properties[], which specifies the default Leader Pinning configuration for all topics that don’t have `redpanda.leaders.preference` explicitly set. It accepts values in the same format as `redpanda.leaders.preference`. Default: `none`

Reviewer 2: ```suggestion
* Set the cluster configuration property config_ref:default_leaders_preference,true,properties/cluster-properties[], which specifies the default Leader Pinning configuration for all topics that don’t have `redpanda.leaders.preference` explicitly set. It accepts values in the same format as `redpanda.leaders.preference`, where the default is `none`.
```

+
[,bash]
----
rpk cluster config set default_leaders_preference ordered_racks:<rack1>,<rack2>
----

If there is more than one broker in the preferred AZ (or AZs), Leader Pinning distributes partition leaders uniformly across brokers in the AZ.

endif::[]

ifdef::env-cloud[]
== Set leader rack preferences

Configure Leader Pinning if you have Redpanda deployed in a multi-AZ or multi-region cluster and your ingress is concentrated in a particular AZ or region.

Use the topic configuration property `redpanda.leaders.preference` to configure Leader Pinning for individual topics. The property accepts the following string values:

** `none`: Disable Leader Pinning for the topic.
** `racks:<rack1>[,<rack2>,...]`: Specify the preferred location (rack) of all topic partition leaders. The list can contain one or more racks, and you can list the racks in any order. Spaces in the list are ignored, for example: `racks:rack1,rack2` and `racks: rack1,  rack2` are equivalent. You cannot specify empty racks, for example: `racks: rack1,,rack2`. If you specify multiple racks, Redpanda tries to distribute the partition leader locations equally across brokers in these racks.

> ** `ordered_racks:<rack1>[,<rack2>,...]`: (Supported in Redpanda version 26.1 or later) Specify the preferred racks in priority order. Redpanda places leaders in the first listed rack when available, failing over to each subsequent rack when higher-priority racks are unavailable. If all listed racks are unavailable, leaders fall back to any other available brokers. Brokers with no rack assignment are treated as lowest priority.

Reviewer 2: ```suggestion
** `ordered_racks:<rack1>[,<rack2>,...]`: Supported in Redpanda v26.1 or later. Specify the preferred racks in priority order. Redpanda places leaders in the first listed rack when available, failing over to each subsequent rack when higher-priority racks are unavailable. If all listed racks are unavailable, leaders fall back to any other available brokers. Brokers with no rack assignment are treated as lowest priority.
```

To find the rack identifiers of all brokers, run:

[,bash]
----
rpk cluster info
----

> 

Reviewer 2: output?

Reviewer: added
